### PR TITLE
Proposal: multitask and transfer learning

### PR DESCRIPTION
**Pull request by deontologician**
Create a benchmark for multi task learning (beginning with performance across Atari games), then develop a reinforcement learning architecture that beats existing techniques on the benchmark. The benchmark will be able to measure both absolute performance on all tasks (e.g. top scores on all Atari games) and also how much more quickly a given architecture learns an unseen task when it is first trained on unrelated tasks (e.g. ratio of score trained vs. untrained on a new game and a fixed number of frames).

[Rendered view of the proposal](https://github.com/deontologician/ai-on.org/blob/proposal-mtatl/projects/multitask-and-transfer-learning.md)

### CODE REVIEW

**Comment by fchollet** [comment]
@delip @0bserver07 Would you like to review this project proposal?

**Comment by 0bserver07** [comment]
Well formatted, I love this topic, I'm currently going after a similar research project for lab-scientific environments with limited supervision : )!

In regards to the question that the project is going after, this might be either too broad or too narrow.

In terms of creating only a benchmark, using already existing benchmarks to start from would be wiser, then moving on to try new hybrid architectures on the environments, since after all this is directed towards `transfer learning` & `multi-task learning`.

For example, currently, the environments provided for ([world_of_bits](https://universe.openai.com/envs#world_of_bits)) soon tend to have benchmark released for transfer learning, ( [as mentioned here](https://openai.com/blog/universe/) ). Using something like this as a starting point could be more productive. 

Here is another relevant work from Berkley RL lab, [Benchmarking Deep Reinforcement Learning for Continuous Control](https://arxiv.org/abs/1604.06778), which contains various benchmarks that can be built on top.

Now in terms of the transfer learning, I looked at the given related work, those are good, however to answer the transfer learning approach, there have been more solid ground work in that area to look at first (which could be directly applied for questions proposed as research):

1. [Reinforcement Learning with Unsupervised Auxiliary Tasks](https://arxiv.org/abs/1611.05397v1), introduces the (UNREAL) agent to combine main agent and auxiliary tasks learned (which could be a shared task for several environments)

2. [Learning to Navigate in Complex Environments](https://arxiv.org/abs/1611.03673v2), introduces (The online multi learner algorithm for multi task learning), which is a neural network-based agents that have modular structures and that are trained on multiple tasks, with inputs coming from different modalities (vision, depth, past rewards and past
actions).

3. Meanwhile on a similar track, but for real world environments, here is another paper mentioning a semi-supervised approach to RL agents, in favor of generalizing skills that used / happens in most environments [Generalizing Skills with Semi-Supervised Reinforcement Learning](https://arxiv.org/abs/1612.00429v1). On Page 2, Figure, there is a comparison how:
> *while in transfer learning, the policy is finetuned using a known reward function in the unlabeled MDP set. Semi-supervised RL is distinct in that it involves using experience from the unlabeled set without access to the reward function.*

and then, the authors point out:

> *Note, however, that the method that we develop could be used for transfer learning problems where the state and reward are consistent
across domains.*

4. More on the robotics and real world transfer learning, this paper suggests few other ways to learn a maze, other than having access to navigational information and path, [Deep Reinforcement Learning with Successor Features for Navigation across Similar Environments](https://arxiv.org/abs/1612.05533v1), similar approaches can be introduced to specific (limited) environments. 

5. bonus: Transfer learning can be introduced through different fashions, not just as actor critic method style. The notion of having multi-agents exploring new areas as for transfer learning seem to be a more feasible approach to be expanded on. In this regards, [ Learning to Communicate with Deep Multi-Agent Reinforcement Learning](https://arxiv.org/pdf/1605.06676v2.pdf), introduces how agents and communicate to improves rewards functions when one of them might taking a leap.


The topic is interesting, lots of active research based results are coming out, but for this proposal, I would suggest leaning more towards one way more, since it can be too broad to go after replicating / doing all those tasks in one research.

The applied research version of this would be creating an RL-Agent that reuses one of those mentioned state-of-art benchmark papers and improve up-on it.

**Comment by deontologician** [comment]
Thanks for taking the time to make such a thorough response, I really appreciate it!

Ok, so re:other benchmarks. The biggest issue I have with existing benchmarks, for this purpose, is that they measure absolute reward on one game, not how much knowledge is transferred between games. Individual papers will measure it, but they do it in effectively ad-hoc ways that are hard to compare across papers. The kind of benchmark I want to make is intrinsically about doing cross-validation after training on different subsets of environments. That's not something I've seen done yet, but I think it's the best approach to measuring transfer, since there tend not to be a lot of different kinds of RL environments available (OpenAI is definitely improving this situation). So the benchmark is more about having a nice baseline for how to measure progress across games with wildly varying amounts of absolute reward. Ultimately, I don't think it'll take too long to build, since the hard parts would be coding the environments themselves and I'm definitely reusing those.

Also, as a side-note I think it's methodologically important to specify up front what the measuring stick is, before coming up with architectures.

As for architectures themselves, I have a bunch of ideas. I think the recent paper [Loss is its own reward](https://arxiv.org/abs/1612.07307) and [predictron](https://arxiv.org/abs/1612.08810) both point the way towards solving the sparse reward issues in RL by coming up with auxiliary rewards that incidentally build representations that are reuseable. But I don't want to flesh the ideas out too much until the benchmark is in place for the aforementioned reasons.

**Comment by 0bserver07** [comment]
Sounds good! Since it's in the Fundamental research already, there can be room to wiggle the problem around.
Maybe update the information to include parts of the last message you posted, as in:
>  Individual papers will measure it, but they do it in effectively ad-hoc ways that are hard to compare across papers. The kind of benchmark I want to make is intrinsically about doing cross-validation after training on different subsets of environments.

and

>Also, as a side-note I think it's methodologically important to specify up front what the measuring stick is, before coming up with architectures.

The `Predictron` is pretty good, I'm reviewing that paper this weekend!

I'm guessing as a starting point:
* Determining the environments
* Benchmarks to look for
and lastly including the results / or related papers to the chosen environments would be cool.

I'm happy to help and find the relevant work in that regard and try getting the benchmarks.

**Review by 0bserver07 — APPROVED** [review]

**Comment by deontologician** [comment]
Ok cool, I've updated the proposal a bit, hopefully made it clearer about the goals, and I've added the UNREAL paper to the bottom though I haven't had time to add the summary. I'll be digging into all of the papers you suggested and I'll add them as I go

**Comment by fchollet** [comment]
LGTM, merging. Thanks @0bserver07 for reviewing!

**Comment by 0bserver07** [comment]
Absolutely :)!

**Comment by fchollet** [comment]
Please take note of the formatting changes I did: https://github.com/AI-ON/ai-on.org/commit/dc3fb4b48af5127b30409e5467fd6414284ddca3
